src/collision.c

```diff
@@ -104,6 +172,192 @@ int CollideSprite( const glTexture* at, const int asx, const int asy, const Vect
 }
 
 
+/**
+ * @brief Checks whether or not a sprite collides with a polygon.
+ *
+ *    @param[in] at Polygon a.
+ *    @param[in] ap Position in space of polygon a.
+ *    @param[in] bt Texture b.
+ *    @param[in] bsx Position of x of sprite b.
+ *    @param[in] bsy Position of y of sprite b.
+ *    @param[in] bp Position in space of sprite b.
+ *    @param[out] crash Actual position of the collision (only set on collision).
+ *    @return 1 on collision, 0 else.
+ */
+int CollideSpritePolygon( const CollPoly* at, const Vector2d* ap,
+      const glTexture* bt, const int bsx, const int bsy, const Vector2d* bp,
+      Vector2d* crash )
+{
+   int x,y;
+   int ax1,ax2, ay1,ay2;
+   int bx1,bx2, by1,by2;
+   int inter_x0, inter_x1, inter_y0, inter_y1;
+   int rbsy;
+   int bbx, bby;
+
+#if DEBUGGING
+   /* Make sure the surfaces have transparency maps. */
+   if (bt->trans == NULL) {
+      WARN(_("Texture '%s' has no transparency map"), bt->name);
+      return 0;
+   }
+#endif /* DEBUGGING */
+
+   /* a - cube coordinates */
+   ax1 = (int)VX(*ap) + (int)(at->xmin);
+   ay1 = (int)VY(*ap) + (int)(at->ymin);
+   ax2 = (int)VX(*ap) + (int)(at->xmax);
+   ay2 = (int)VY(*ap) + (int)(at->ymax);
+
+   /* b - cube coordinates */
+   bx1 = (int)VX(*bp) - (int)(bt->sw)/2;
+   by1 = (int)VY(*bp) - (int)(bt->sh)/2;
+   bx2 = bx1 + bt->sw - 1;
+   by2 = by1 + bt->sh - 1;
+
+   /* check if bounding boxes intersect */
+   if((bx2 < ax1) || (ax2 < bx1)) return 0;
+   if((by2 < ay1) || (ay2 < by1)) return 0;
+
+   /* define the remaining binding box */
+   inter_x0 = MAX( ax1, bx1 );
+   inter_x1 = MIN( ax2, bx2 );
+   inter_y0 = MAX( ay1, by1 );
+   inter_y1 = MIN( ay2, by2 );
+
+   /* real vertical sprite value (flipped) */
+   rbsy = bt->sy - bsy - 1;
+
+   /* set up the base points */
+   bbx =  bsx*(int)(bt->sw) - bx1;
+   bby = rbsy*(int)(bt->sh) - by1;
+   for (y=inter_y0; y<=inter_y1; y++) {
+      for (x=inter_x0; x<=inter_x1; x++) {
+         /* compute offsets for surface before pass to TransparentPixel test */
+         if ((!gl_isTrans(bt, bbx + x, bby + y))) {
+            if (pointInPolygon( at, ap, (float)x, (float)y )) {
+               crash->x = x;
+               crash->y = y;
+               return 1;
+            }
+         }
+      }
+   }
+
+   return 0;
+}
+
+
+/**
+ * @brief Checks whether or not two polygons collide.
+ *  /!\ The function is not symmetric: the points of polygon 2 are tested
```

Comment:
Not sure if it would be worth to make it always swap the polygons so that polygon 2 is the smaller one or have a wrapper function do that.